pilot/cmd/pilot-agent/status/server.go

```diff
@@ -502,8 +493,68 @@ func negotiateMetricsFormat(contentType string) expfmt.Format {
 	return expfmt.FmtText
 }
 
-func processMetrics(metrics []byte) []byte {
-	return bytes.ReplaceAll(metrics, []byte("\n\n"), []byte("\n"))
+// metricReader used to remove all the blank, "# EOF" or incomplete lines in envoy metrics.
+// It makes the envoy metric compatible with FmtOpenMetrics (https://github.com/istio/istio/pull/33550)
+type metricReader struct {
+	reader  io.ReadCloser
+	buf     *bufio.Reader
+	readBuf *bytes.Buffer
+}
+
+func NewMetricReader(reader io.ReadCloser) *metricReader {
+	return &metricReader{reader: reader, buf: bufio.NewReader(reader)}
+}
+
+// WriteTo io.copy will call this function first.
+// It will drop every blank line and incomplete line
+func (r *metricReader) WriteTo(w io.Writer) (n int64, err error) {
+	var line []byte
+	var isEmptyLine, isBufferFull bool
+	newLine, EOFLine := []byte{'\n'}, []byte("# EOF")
+
+	for {
+		line, err = r.buf.ReadSlice(newLine[0])
+		// Once get unexpected error, drop last line that may be incomplete
+		if err != nil && err != io.EOF && err != bufio.ErrBufferFull {
+			break
+		}
+		// Remove "# EOF" avoid terminates the full exposition
+		if bytes.HasPrefix(line, EOFLine) {
+			line = line[len(EOFLine):]
+		}
+		// If last line size equal buffer size, need to add "\n"
+		isEmptyLine = bytes.Equal(line, newLine) && !isBufferFull
+		if isEmptyLine {
+			continue
+		}
+		wLength, werr := w.Write(line)
+		n += int64(wLength)
+
+		isBufferFull = false
+		if werr == nil && err == bufio.ErrBufferFull {
+			isBufferFull = true
```

Comment:
If there is a very large line (larger than buffer size), it will lead to wrong line feed.Although bufio.reader get 4k buffer, I still worry about it(current, our line is about 1.1k in size).

Once line length over the buffer size, ReadSlice while return a bufio.ErrBufferFull error and not '\n' in line.So we need to avoid exiting functions early because of errors. And add "\n" if last line size equal buffer size.